Instant Search: improve accessibility of close button

bluefuton opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Make sure the overlay close button is a real `<button>` to support the associated keyboard events.
* Add a focus outline style for the close button.
* Make sure the button is appropriately labelled for screen readers.
* Make sure the SVG is not focusable.

References:
- https://www.sarasoueidan.com/blog/accessible-icon-buttons/ (used technique #3 here)
- https://codepen.io/andybelldesign/pen/Vxpjvo (button style reset)

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
Test the overlay in a selection of themes and check that the visual appearance of the close button has not changed.

Inspect the button with your browser's accessibility tools and ensure that you can see the correct label:

<img width="854" alt="Screen Shot 2020-09-29 at 16 35 31" src="https://user-images.githubusercontent.com/17325/94509888-e0bd6280-0271-11eb-824f-d077d4adc353.png">

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
Not required.

bluefuton committed:
Instant Search: improve a11y of close button

bluefuton commented on modules/search/instant-search/components/overlay.scss:
```diff
@@ -1,4 +1,4 @@
-$overlay-vertical-padding: 3.5em;
+$overlay-vertical-padding: 4em;
```
This is to give enough space for the outline styles without overlapping the search box.

<img width="1089" alt="Screen Shot 2020-09-29 at 16 40 18" src="https://user-images.githubusercontent.com/17325/94510131-7c4ed300-0272-11eb-8f16-369d33d24111.png">

bluefuton commented on modules/search/instant-search/components/search-results.jsx:
```diff
@@ -166,15 +166,15 @@ class SearchResults extends Component {
 				className="jetpack-instant-search__search-results"
 			>
-				<a
+				<button
```
An `a` with `role="button"` does not provide all the same keyboard events, like being able to activate the button using either the ENTER or SPACE keys. We could reimplement those ourselves, but it makes sense to just use the `button` element if we can.

bluefuton committed:
Use dotted outline

jsnmoon commented on modules/search/instant-search/components/search-results.scss:
```diff
@@ -121,3 +129,18 @@
+    margin: 0;
+    text-decoration: none;
+    cursor: pointer;
+    -webkit-appearance: none;
+    -moz-appearance: none;
```
Could we use [`appearance: none;`](https://developer.mozilla.org/en-US/docs/Web/CSS/appearance) here? The build process will add these vendor prefixes as necessary.

jsnmoon commented on modules/search/instant-search/components/search-results.jsx:
```diff
@@ -166,15 +166,15 @@ class SearchResults extends Component {
 					onKeyPress={ this.onKeyPressHandler }
-					role="button"
 					tabIndex="0"
+					aria-label={ __( 'Close search results', 'jetpack' ) }
```
Wouldn't this duplicate the aria-label already provided by the Gridicon? Or is that expected/fine?

<img width="485" alt="Screen Shot 2020-10-06 at 3 40 26 PM" src="https://user-images.githubusercontent.com/4044428/95262922-42d61480-07ea-11eb-8391-7fa3e2b672a3.png">

bluefuton replied:
Ah - it's the title on the SVG. I've added an `aria-hidden` on the SVG to address this 👍

jsnmoon reviewed (commented):
Looks good overall. I left a question and a suggestion inline.

bluefuton committed:
Address review feedback

bluefuton committed:
Only turn off focusable for this button

jsnmoon reviewed (changes requested):
`aria-hidden` changes look good, but I discovered another strange behavior: hoving over the close button nudges it down and to the right:


![Oct-07-2020 17-17-31](https://user-images.githubusercontent.com/4044428/95397660-04f7ef80-08c1-11eb-8695-4fb40e5a2ccb.gif)

It looks like we have a lot of styles being applied to `.jetpack-instant-search__overlay-close` from both `modules/search/instant-search/components/search-results.scss` and `modules/search/instant-search/components/overlay.scss`. Perhaps consolidating these styles will help us avoid these styling bugs?

bluefuton commented:
> hoving over the close button nudges it down and to the right:

How curious. What theme and what browser is it @jsnmoon? I'm not seeing it in TwentyTwenty on Safari, FF or Chrome.

Good call on consolidating the styles. Possibly a case for splitting close button into its own component?

jsnmoon commented:
> > hoving over the close button nudges it down and to the right:
> 
> How curious. What theme and what browser is it @jsnmoon? I'm not seeing it in TwentyTwenty on Safari, FF or Chrome.
> 
> Good call on consolidating the styles. Possibly a case for splitting close button into its own component?

Strange, I'm no longer able to reproduce. I was using TwentyTwenty with Chrome 85 when I took the screencap.

jsnmoon commented:
Due to the changes we've introduced to the Gridicon block, we'll also need to update `modules/search/instant-search/components/test/notice.test.js` to include `aria-hidden` and `focusable` attributes. Travis is failing due to this discrepancy.

bluefuton committed:
Update test snapshot

bluefuton committed:
Consolidate styles

bluefuton commented:
> Due to the changes we've introduced to the Gridicon block, we'll also need to update modules/search/instant-search/components/test/notice.test.js to include aria-hidden and focusable attributes

Good spot! I've updated the Jest snapshot. I used:

`./node_modules/jest/bin/jest.js modules/search/instant-search --updateSnapshot`

jsnmoon reviewed (approved):
🚢
